cli/azd/cmd/init.go

```diff
@@ -69,6 +71,128 @@ func (i *initAction) SetupFlags(
+			progress.Message("Step 1")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 2")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 3")
+			time.Sleep(1 * time.Second)
+			progress.Info("Some additional info")
+
+			return nil
+		})).
+		AddStep(ux.NewProgressStep("Doing thing 3", "Initializing", func(ctx context.Context, stepCtx ux.StepContext, progress *ux.Progress) error {
+			progress.Message("Step 1")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 2")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 3")
+			time.Sleep(1 * time.Second)
+			progress.Warn("I need to warn you about something")
+
+			return nil
+		})).
+		AddStep(ux.NewProgressStep("Doing thing 4", "Initializing", func(ctx context.Context, stepCtx ux.StepContext, progress *ux.Progress) error {
+			progress.Message("Step 1")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 2")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 3")
+			time.Sleep(1 * time.Second)
+			progress.Skip("Don't need to do this")
+
+			return nil
+		})).
+		AddStep(ux.NewInteractiveStep(func(ctx context.Context, stepCtx ux.StepContext) error {
+			console := input.GetConsole(ctx)
+			console.Message(ctx, "This message is written from inside a step.")
+			console.Message(ctx, "")
+
+			console.Prompt(ctx, input.ConsoleOptions{
+				Message:      "Enter a value",
+				DefaultValue: "Default",
+				Help:         "help",
+			})
+
+			console.Select(ctx, input.ConsoleOptions{
+				Message:      "Select an option from the list",
+				Options:      []string{"Option 1", "Option 2", "Option 3"},
+				DefaultValue: "Option 1",
+				Help:         "help",
+			})
+
+			userValue, _ := console.Confirm(ctx,
+				input.ConsoleOptions{
+					Message:      "Should next step fail?",
+					DefaultValue: false,
+					Help:         "help",
+				})
+
+			stepCtx.SetValue("fail", userValue)
+			return nil
+		})).
+		AddStep(ux.NewProgressStep("Doing thing 5", "Initializing", func(ctx context.Context, stepCtx ux.StepContext, progress *ux.Progress) error {
+			progress.Message("Step 1")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 2")
+			time.Sleep(1 * time.Second)
+			progress.Message("Step 3")
+			time.Sleep(1 * time.Second)
+
+			failValue, ok := stepCtx.GetValue("fail")
+			if !ok {
+				return errors.New("Cannot find `fail` value in step context")
+			}
+
+			if failValue.(bool) {
+				return errors.New("Something bad happened and cannot continue")
+			}
+
+			// Set some result
+			stepCtx.SetValue("result", "More information to display in the complete func")
+
+			return nil
+		})).
+		Complete("Finished the thing!", func(ctx context.Context, stepCtx ux.StepContext) error {
+			result, ok := stepCtx.GetValue("result")
+			if ok {
+				fmt.Println(result.(string))
+			}
+
+			return nil
+		}).
+		Error("Initialization failed!", func(ctx context.Context, stepCtx ux.StepContext, err error) {
+			fmt.Println(err.Error())
+		}).
+		Run(ctx)
```

Comment:
Would this design force us to declare all the flow in the `cmd` command using this builder?

I am wondering about things like:
- Running multiple steps at the same time
- Having some steps depending on others (like step-bifurcation if-else)
- Sharing data between steps (globals, or step-output as step-input)

It feels like we would be creating/describing something similar to an azdo pipeline or github flow.  Sequential steps, conditions and variables. I don't feel very happy about it.

I like the `*ux.Progress` design.  
Can we just pass that component thru the flow and add some functionality to it like
- AddStep()

Then, instead of declaring the steps first, we can let the flow to decide when a new step should be created.  Then, when calling `AddStep()` from a current step, we can dynamically handle `sub-steps` to add more indentation. And that's set directly inside the command implementation (from the controller directly, thinking as MVC)

Reply:
Or, in case we want to use this builder, can we make controllers to be the ones who defines the steps and flow?

So, here, instead of using the builder to create the steps for `init`, we would call something like `initManager.GetSteps()` which would delegate `creating the Action Printer` to the actual controller, and not to the View

Reply:
I imagine everything won't/shouldn't use the builder and steps could be added incrementally through our business logic and then scheduled within the commands.  

We can then introduce like a nested step or composite step and other step types that handle multiple sub steps running concurrently.